core/services/pipeline/orm.go

```diff
@@ -50,6 +59,136 @@ func (o *orm) CreateSpec(ctx context.Context, tx *gorm.DB, pipeline Pipeline, ma
 	return spec.ID, errors.WithStack(err)
 }
 
+func (o *orm) CreateRun(db *gorm.DB, run *Run) (err error) {
+	if run.CreatedAt.IsZero() {
+		return errors.New("run.CreatedAt must be set")
+	}
+	if err = db.Create(run).Error; err != nil {
+		return errors.Wrap(err, "error inserting pipeline_run")
+	}
+	return err
+}
+
+// StoreRun will persist a partially executed run before suspending, or finish a run.
+func (o *orm) StoreRun(db *sql.DB, run *Run) (bool, error) {
+	finished := run.FinishedAt.Valid
+	var restart bool
```

Comment:
It is unclear to me what restart does/controls

Reply:
If new data came in before StoreRun was called, we're able to immediately update the data and re-execute. This is handled inside `runner.Run()`. I'll update the method comment here